frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java

```diff
@@ -0,0 +1,144 @@
+    val branch = branchName != null
+        ? getManagedBranchByName(anActionEvent, branchName)
+        : null;
+
+    if (branch == null) {
+      presentation.setEnabled(false);
+      presentation.setDescription(getNonHtmlString("action.GitMachete.description.disabled.undefined.machete-branch")
+          .fmt("Rename", getQuotedStringOrCurrent(branchName)));
+    } else {
+      presentation.setDescription(
+          getNonHtmlString("action.GitMachete.BaseRenameAction.description").fmt(branch.getName()));
+    }
+  }
+
+  @Override
+  @ContinuesInBackground
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    val gitRepository = getSelectedGitRepository(anActionEvent);
+    val branchName = getNameOfBranchUnderAction(anActionEvent);
+    val branch = branchName != null ? getManagedBranchByName(anActionEvent, branchName) : null;
+    val branchLayout = getBranchLayout(anActionEvent);
+    val graphTable = getGraphTable(anActionEvent);
+
+    if (gitRepository == null || branchName == null || branchLayout == null || branch == null || graphTable == null) {
+      return;
+    }
+
+    rename(gitRepository, graphTable, branch, branchLayout);
+  }
+
+  @ContinuesInBackground
+  @IgnoreUIThreadUnsafeCalls("com.virtuslab.gitmachete.frontend.actions.common.BranchCreationUtils.waitForCreationOfLocalBranch"
+      + "(git4idea.repo.GitRepository, java.lang.String)")
+  @UIEffect
+  private void rename(GitRepository gitRepository, BaseEnhancedGraphTable graphTable, IManagedBranchSnapshot branch,
+      BranchLayout branchLayout) {
+    val project = gitRepository.getProject();
+
+    val gitNewBranchDialog = new GitNewBranchDialog(project, Collections.singletonList(gitRepository),
+        getString("action.GitMachete.BaseRenameAction.description").fmt(branch.getName()),
+        branch.getName(),
+        /* showCheckOutOption */ false,
+        /* showResetOption */ false,
+        /* showSetTrackingOption */ false,
+        /* localConflictsAllowed */ false,
+        GitBranchOperationType.RENAME);
+
+    val options = gitNewBranchDialog.showAndGetOptions();
+
+    if (options != null) {
+
+      Path macheteFilePath = gitRepository.getMacheteFilePath();
+
+      val newBranchLayout = branchLayout.rename(branch.getName(), options.getName());
+      val branchLayoutWriter = ApplicationManager.getApplication().getService(IBranchLayoutWriter.class);
+      val gitBrancher = GitBrancher.getInstance(project);
+
+      new Task.Backgroundable(gitRepository.getProject(), getString("action.GitMachete.BaseRenameAction.task-title")) {
+        @Override
+        @UIThreadUnsafe
+        @ContinuesInBackground
+        public void run(ProgressIndicator indicator) {
+          graphTable.disableEnqueuingUpdates();
+          gitBrancher.renameBranch(branch.getName(), options.getName(), Collections.singletonList(gitRepository));
+
+          runWriteActionOnUIThread(() -> {
```

Comment:
Nit: I'd even use
```suggestion
          blockingRunWriteActionOnUIThread(() -> {
```

Shouldn't harm performance in any significant way, and should protect against some crazy race conditions (like `waitForCreationOfLocalBranch` finishing before `writeBranchLayout` is complete)